lib/sycamore/sycamore/writers/duckdb_writer.py

```diff
@@ -0,0 +1,145 @@
+        )
+        assert _narrow_list_of_doc_records(records), f"Found a bad record in {records}"
+        assert isinstance(target_params, DuckDBTargetParams)
+
+        headers = ["uuid", "embeddings", "properties", "text_representation", "bbox", "shingles", "type"]
+        creation = True
+        # schema = pa.schema(
+        #     [
+        #         ("uuid", pa.string()),
+        #         ("embeddings", pa.list_(pa.float32())),
+        #         ("properties", pa.map_(pa.string(), pa.string())),
+        #         ("text_representation", pa.string()),
+        #         ("bbox", pa.list_(pa.float32())),
+        #         ("shingles", pa.list_(pa.int64())),
+        #         ("type", pa.string()),
+        #     ]
+        # )
+
+        def write_batch(batch_data: dict):
+            pa_table = pa.Table.from_pydict(batch_data)  # noqa
+            client = duckdb.connect(str(dict_params.get("db_url")))
+            nonlocal creation
+            if creation:
+                client.sql(f"CREATE TABLE {dict_params.get('table_name')} AS SELECT * FROM pa_table")
+                creation = False
+            else:
+                client.sql(f"INSERT INTO {dict_params.get('table_name')} SELECT * FROM pa_table")
+            for key in batch_data:
+                batch_data[key].clear()
+
+        batch_data: dict[str, list[Any]] = {key: [] for key in headers}
+
+        for r in records:
+            # Append the new data to the batch
+            batch_data["uuid"].append(r.uuid)
+            batch_data["embeddings"].append(r.embeddings)
+            batch_data["properties"].append(r.properties)
+            batch_data["text_representation"].append(r.text_representation)
+            batch_data["bbox"].append(r.bbox)
+            batch_data["shingles"].append(r.shingles)
+            batch_data["type"].append(r.type)
+
+            # If we've reached the batch size, write to the database
+            if batch_data.__sizeof__() >= N:
+                batch_data["properties"] = (
+                    [convert_to_str_dict(i) for i in batch_data["properties"]] if batch_data["properties"] else []
+                )
+                write_batch(batch_data)
+        # Write any remaining records
+        if len(batch_data["uuid"]) > 0:
+            write_batch(batch_data)
+
+    def create_target_idempotent(self, target_params: BaseDBWriter.TargetParams):
+        assert isinstance(target_params, DuckDBTargetParams)
+
+    def get_existing_target_params(self, target_params: BaseDBWriter.TargetParams) -> "DuckDBTargetParams":
+        assert isinstance(target_params, DuckDBTargetParams)
+        return DuckDBTargetParams(db_url=target_params.db_url, table_name=target_params.table_name)
+
+
+@dataclass
+class DuckDBDocumentRecord(BaseDBWriter.Record):
+    uuid: str
+    embeddings: Optional[list[float]] = None
+    properties: Optional[dict[str, Any]] = None
```

Comment:
maybe make this the dict[str, str] and do your conversion here

Reply:
Discussed, would be better doing this later